perllib/FixMyStreet/App/Controller/Waste/Whitespace.pm

```diff
@@ -20,22 +22,96 @@ has log_ident => (
     }
 );
 
+has whitespace_config => (
+    is => 'rw',
+    isa => 'HashRef',
+    default => sub { {} }
+);
+
 sub receive_whitespace_event_notification : Path('/waste/whitespace') : Args(0) {
     my ($self, $c) = @_;
 
+    $c->detach('soap_error', [ 'Invalid method', 405 ]) unless $c->req->method eq 'POST';
+
+    $self->whitespace_config(FixMyStreet->config('COBRAND_FEATURES')->{whitespace}{bexley});
+
     my %headers = $c->req->headers->flatten;
     $self->log($c->req->method);
     $self->log(\%headers);
     $self->log($c->req->parameters);
     if ($c->req->body) {
         my $soap = join('', $c->req->body->getlines);
         $self->log($soap);
+        my $env = SOAP::Deserializer->deserialize($soap);
+        if ($env->valueof('//WorksheetPoke/secret') ne $self->whitespace_config->{missed_collection_secret}) {
+            return $c->detach('soap_error', [ 'Unauthorized', 401 ]);
+        };
+        my %params;
+        $params{external_id} = $env->valueof('//WorksheetPoke/worksheetId');
+        $params{status} = $env->valueof('//WorksheetPoke/status');
+        $params{completed_dt} = $env->valueof('//WorksheetPoke/completedDate');
+        $params{update_ref} = $env->valueof('//WorksheetPoke/status');
+        return $c->detach('soap_error', [ 'Bad request', 400 ]) if _check_params(\%params);
+        $self->update_report(\%params);
     } else {
         $self->log('No body');
+        return $c->detach('soap_error', [ 'Bad request', 400 ]);
     }
 
     $c->response->status(200);
     $c->response->body('OK');
 }
 
+sub update_report {
+    my ($self, $params) = @_;
+
+    my $report = FixMyStreet::DB->resultset("Problem")->search( { external_id => 'Whitespace-' . $params->{external_id} } )->first;
+    return unless $report;
+
+    my $status_hash = $self->whitespace_config->{missed_collection_state_mapping}{$params->{status}};
```

Comment:
Might be worth checking the status is actually found in the state_mapping, in case they add new statuses in the future or similar.

Reply:
If the status isn't present the data should now be rejected

Reply:
It appears the statuses in the push are actually only "Completed" or "Cancelled", not the states you'd get from the worksheet details.